### PR TITLE
Socketry::Resolver: pluggable DNS resolution support with timeouts

### PR DESCRIPTION
**Pull request by tarcieri**
This provides an abstract API for mapping IP addresses and domain names to `Resolv::IPv4` and `Resolv::IPv6` addresses.

It provides two backends:

* `Socketry::Resolver::System`: based on `IPSocket.getaddress`, which is implemented using the POSIX `getaddrinfo(3)` function, with timeouts provided by Ruby's `Timeout` module.
* `Socketry::Resolver::Resolv`: based on `Resolv::DNS`, a pure Ruby implementation of the DNS protocol, which natively implements timeouts using Ruby's asynchronous I/O primitives.

I have made `Socketry::Resolver::System` the default. [I am aware that Ruby's Timeout module is broken](http://blog.headius.com/2008/02/rubys-threadraise-threadkill-timeoutrb.html) in a multithreaded context (that is, of course, the whole point of this library existing in the first place), but I think it will be a generally less problematic default, as I have experienced a long history of issues with non-system resolvers.

cc @indirect @ioquatix @digitalextremist @zanker @ixti @TiagoCardoso1983

### CODE REVIEW

**Comment by indirect** [comment]
this looks great to me 👍

File: lib/socketry/exceptions.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module Socketry
+  # Generic catch all for all Socketry errors
+  SocketError = Class.new(IOError)
```
Comment (ixti) [inline lib/socketry/exceptions.rb]: - why not simply `Error`? IMHO `Socketry::Error` looks better than `Socketry::SocketError`.
- why `MyClass = Class.new(ParentClass)`? Why not `class MyClass < ParentClass; end`?

Comment (ixti) [inline lib/socketry/exceptions.rb]: IMHO if one goes `MyClass = Class.new(ParentClass)`, he must also use that style everywhere else:

``` ruby
module Socketry
  Resolver = Module.new do
    Error = Class.new(AddressError)
  end
end
```

Comment (ixti) [inline lib/socketry/exceptions.rb]: althought it's not a valid syntax of course ;))

**Comment by ixti** [comment]
awesome! :+1:

File: lib/socketry/exceptions.rb (hunk 1 above)
Comment (tarcieri) [inline lib/socketry/exceptions.rb]: +1 on the first note.

Re: the `Class.new` syntax, personal preference. I find it less ugly than using a `; end`. YMMV.

Comment (ixti) [inline lib/socketry/exceptions.rb]: Mmm... I can see how `X = Class.new(Y)` might look better than `class X < Y; end` but I think it makes code less consistent :D Anyway, I have no strong feeling one way or the other ;))

**Comment by ioquatix** [comment]
Not sure if this is relevant here, but have you considered using the celluloid DNS resolver which is fully functional, tested, etc?

**Comment by tarcieri** [comment]
@ioquatix the goal is to provide pluggable backends that can be used depending on your needs. I would like to write one that uses the `Celluloid::IO` DNS resolver, particularly when Socketry is used with `Celluloid::IO`, but I also don't want to make Celluloid a hard dependency of Socketry.

**Comment by ixti** [comment]
@tarcieri I think backend can be shipped as part of Socketry, but not required by default (or even magically required if `Celluloid::IO` is defined. Thus we will need Celluloid::IO as tests dependency and at the same time backend will stay close to core :D

File: lib/socketry/resolver.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module Socketry
+  # DNS resolution subsystem
+  module Resolver
+    module_function
+
+    def addr(address)
+      return ::Resolv::IPv4.create(address)
```
Comment (TiagoCardoso1983) [inline lib/socketry/resolver.rb]: Isn't there a way to unify these IPvX calls without catching exceptions? The `ipaddr` stdlib could provide an instance which responds to `#ipv4?` and `#ipv6?`.

Comment (tarcieri) [inline lib/socketry/resolver.rb]: Nice, I like that idea.

**Pull request merged**